src/smc-webapp/frame-editors/latex-editor/pythontex.ts

```diff
@@ -0,0 +1,102 @@
+/*
+Run PythonTeX
+*/
+
+import { exec, ExecOutput } from "../generic/client";
+import { parse_path } from "../frame-tree/util";
+import { ProcessedLatexLog, Error } from "./latex-log-parser";
+import { BuildLog } from "./actions";
+
+// command documentation
+//
+// we limit the number of jobs, could be bad for memory usage causing OOM or whatnot
+// -j N, --jobs N        Allow N jobs at once; defaults to cpu_count().
+//
+// --rerun={never,modified,errors,warnings,always}
+// This sets the threshold for re-executing code.
+// By default, PythonTEX will rerun code that has been modified or that produced errors on the last run.
+// "always" executes all code always
+
+export async function pythontex(
+  project_id: string,
+  path: string,
+  time: number,
+  force: boolean,
+  status: Function
+): Promise<ExecOutput> {
+  const { base, directory } = parse_path(path);
+  const args = ["--jobs", "2"];
+  if (force) {
+    // forced build implies to run all snippets
+    args.push("--rerun=always");
```

Comment:
for a forced build, we tell pythontex to re-run all snippets unconditionally -- otherwise only those "sessions" (there can be several chains of snippets for execution per file) with modifications run again